Index: kernel-rockchip/arch/arm64/boot/dts/rockchip/rk3588s-12c.dtsi
===================================================================
--- kernel-rockchip.orig/arch/arm64/boot/dts/rockchip/rk3588s-12c.dtsi
+++ kernel-rockchip/arch/arm64/boot/dts/rockchip/rk3588s-12c.dtsi
@@ -428,6 +428,7 @@
   max-frequency = <150000000>;
   no-sdio;
   no-mmc;
+  bus-width = <4>;
   cap-sd-highspeed;
   cap-mmc-highspeed;
  // clock-frequency = <100000000>;
Index: kernel-rockchip/drivers/soc/rockchip/Kconfig
===================================================================
--- kernel-rockchip.orig/drivers/soc/rockchip/Kconfig
+++ kernel-rockchip/drivers/soc/rockchip/Kconfig
@@ -296,4 +296,9 @@ source "drivers/soc/rockchip/minidump/Kc
 
 endmenu
 
+config ROCKCHIP_FYDETAB_GRF
+	bool "initialize grf with specific values on FydeTab"
+	default n
+	help
+	  If unsure, say n.
 endif
Index: kernel-rockchip/drivers/soc/rockchip/grf.c
===================================================================
--- kernel-rockchip.orig/drivers/soc/rockchip/grf.c
+++ kernel-rockchip/drivers/soc/rockchip/grf.c
@@ -229,10 +229,15 @@ static const struct rockchip_grf_info rk
 	.num_values = ARRAY_SIZE(rk3399_defaults),
 };
 
+#define RK3588_SYS_GRF_SOC_CON6 	0x0318
 #define RK3588_SYS_GRF_SOC_CON7		0x031c
 
 static const struct rockchip_grf_value rk3588_sys_grf_defaults[] __initconst = {
 	{ "Connect EDP hpd to IO", RK3588_SYS_GRF_SOC_CON7, HIWORD_UPDATE(0x3, 0x3, 14) },
+#ifdef CONFIG_ROCKCHIP_FYDETAB_GRF
+	/* Without this, MiscroSD card won't work normally with bus width 4 */
+	{ "jtag switching", RK3588_SYS_GRF_SOC_CON6, HIWORD_UPDATE(0x0, 0x1, 14) },
+#endif
 };
 
 static const struct rockchip_grf_info rk3588_sys_grf __initconst = {
