Index: kernel-rockchip/usr/Makefile
===================================================================
--- kernel-rockchip.orig/usr/Makefile
+++ kernel-rockchip/usr/Makefile
@@ -9,13 +9,13 @@
 override size_append := :
 
 compress-y					:= shipped
-compress-$(CONFIG_INITRAMFS_COMPRESSION_GZIP)	:= gzip
-compress-$(CONFIG_INITRAMFS_COMPRESSION_BZIP2)	:= bzip2
-compress-$(CONFIG_INITRAMFS_COMPRESSION_LZMA)	:= lzma
-compress-$(CONFIG_INITRAMFS_COMPRESSION_XZ)	:= xzmisc
-compress-$(CONFIG_INITRAMFS_COMPRESSION_LZO)	:= lzo
-compress-$(CONFIG_INITRAMFS_COMPRESSION_LZ4)	:= lz4
-compress-$(CONFIG_INITRAMFS_COMPRESSION_ZSTD)	:= zstd
+# compress-$(CONFIG_INITRAMFS_COMPRESSION_GZIP)	:= gzip
+# compress-$(CONFIG_INITRAMFS_COMPRESSION_BZIP2)	:= bzip2
+# compress-$(CONFIG_INITRAMFS_COMPRESSION_LZMA)	:= lzma
+# compress-$(CONFIG_INITRAMFS_COMPRESSION_XZ)	:= xzmisc
+# compress-$(CONFIG_INITRAMFS_COMPRESSION_LZO)	:= lzo
+# compress-$(CONFIG_INITRAMFS_COMPRESSION_LZ4)	:= lz4
+# compress-$(CONFIG_INITRAMFS_COMPRESSION_ZSTD)	:= zstd
 
 obj-$(CONFIG_BLK_DEV_INITRD) := initramfs_data.o
 
