Index: kernel-rockchip/arch/arm64/boot/dts/rockchip/rk3588s-orangepi-5.dts
===================================================================
--- kernel-rockchip.orig/arch/arm64/boot/dts/rockchip/rk3588s-orangepi-5.dts
+++ kernel-rockchip/arch/arm64/boot/dts/rockchip/rk3588s-orangepi-5.dts
@@ -119,7 +119,7 @@
 
 &hdmi0 {
 	enable-gpios = <&gpio4 RK_PB6 GPIO_ACTIVE_HIGH>;
-	cec-enable;
+	cec-enable = "true";
 	status = "okay";
 };
 
Index: kernel-rockchip/drivers/gpu/drm/bridge/synopsys/dw-hdmi-qp.c
===================================================================
--- kernel-rockchip.orig/drivers/gpu/drm/bridge/synopsys/dw-hdmi-qp.c
+++ kernel-rockchip/drivers/gpu/drm/bridge/synopsys/dw-hdmi-qp.c
@@ -3265,13 +3265,22 @@ void dw_hdmi_qp_cec_set_hpd(struct dw_hd
 {
 	enum drm_connector_status status = plug_in ?
 		connector_status_connected : connector_status_disconnected;
+  struct edid *edid = NULL;
 
 	if (!hdmi->cec_notifier)
 		return;
 
-	if (!plug_in)
+	if (!plug_in) {
 		cec_notifier_set_phys_addr(hdmi->cec_notifier,
 					   CEC_PHYS_ADDR_INVALID);
+  } else if (hdmi->ddc) {
+    edid = drm_get_edid(&hdmi->connector, hdmi->ddc);
+    if (edid) {
+      if (hdmi->cec_notifier)
+        cec_notifier_set_phys_addr_from_edid(hdmi->cec_notifier, edid);
+      kfree(edid);
+    }
+  }
 
 	if (hdmi->bridge.dev) {
 		if (change && hdmi->cec_adap && hdmi->cec_adap->devnode.registered)
