Index: crosvm/jail/seccomp/aarch64/gpu_common.policy
===================================================================
--- crosvm.orig/jail/seccomp/aarch64/gpu_common.policy
+++ crosvm/jail/seccomp/aarch64/gpu_common.policy
@@ -100,6 +100,7 @@ faccessat: 1
 faccessat2: 1
 getgid: 1
 getegid: 1
+pread64: 1

 ## Rules for vmm-swap
 userfaultfd: 1
Index: crosvm/jail/seccomp/arm/gpu_common.policy
===================================================================
--- crosvm.orig/jail/seccomp/arm/gpu_common.policy
+++ crosvm/jail/seccomp/arm/gpu_common.policy
@@ -98,6 +98,7 @@ geteuid32: 1
 getuid32: 1
 lstat64: 1
 fcntl64: 1
+pread64: 1
 tgkill: 1
 clock_gettime: 1
 clock_gettime64: 1
