Index: kernel-6_12/drivers/usb/typec/mux/pi3dpx1207c.c
===================================================================
--- /dev/null
+++ kernel-6_12/drivers/usb/typec/mux/pi3dpx1207c.c
@@ -0,0 +1,212 @@
+// SPDX-License-Identifier: GPL-2.0-only
+/*
+ * Haoyu pi3dpx1207c RTC driver
+ *
+ * Copyright (C) 2013 MundoReader S.L.
+ * Author: Heiko Stuebner <heiko@sntech.de>
+ *
+ * based on rtc-pi3dpx1207c
+ * Copyright (C) 2010 ROCKCHIP, Inc.
+ */
+
+#include <linux/module.h>
+#include <linux/clk-provider.h>
+#include <linux/i2c.h>
+#include <linux/usb/typec_dp.h>
+#include <linux/usb/typec_mux.h>
+#include <linux/mutex.h>
+#define PI3DPX_BYTES_NUM	9
+
+struct pi3dpx_config {
+    uint8_t full_dp_config[PI3DPX_BYTES_NUM];
+		uint8_t no_dp_config[PI3DPX_BYTES_NUM];
+		uint8_t usb_dp_config[PI3DPX_BYTES_NUM];
+};
+
+struct pi3dpx_priv {
+    struct i2c_client       *client;
+    struct device *dev;
+    unsigned long typec_state;
+    unsigned int new_flip;
+    unsigned int current_flip;
+    struct mutex i2c_rw_lock;
+    struct typec_mux_dev *mux;
+    struct typec_switch_dev *sw;
+};
+
+struct pi3dpx_config redriver_config[2] = {
+	{
+		{0x13 ,0x11 ,0x20 ,0x22 ,0x00 ,0xa7 ,0xa7 ,0xa7 ,0xa7},
+		{0x13 ,0x11 ,0x20 ,0x42 ,0x00 ,0xa7 ,0xa7 ,0xa7 ,0xa7},
+		{0x13 ,0x11 ,0x20 ,0x62 ,0x00 ,0xa7 ,0xa7 ,0xa7 ,0xa7},
+	},
+
+	{
+		{0x13 ,0x11 ,0x20 ,0x22 ,0x00 ,0xa7 ,0xa7 ,0xa7 ,0xa7},
+		{0x13 ,0x11 ,0x20 ,0x52 ,0x00 ,0xa7 ,0xa7 ,0xa7 ,0xa7},
+		{0x13 ,0x11 ,0x20 ,0x82 ,0x00 ,0xa7 ,0xa7 ,0xa7 ,0xa7},
+	}
+
+};
+static int pi3dpx_i2c_write(struct pi3dpx_priv *priv,int bytes, void *src)
+{
+   struct i2c_client *i2c = priv->client;
+   struct i2c_msg msg;
+   int ret;
+   if (bytes != PI3DPX_BYTES_NUM)
+           return -EINVAL;
+   msg.addr = i2c->addr;
+   msg.flags = 0;
+   msg.len = bytes;
+   msg.buf = (u8*)src;
+
+   ret = i2c_transfer(i2c->adapter, &msg, 1);
+   if (ret != 1)
+     dev_err(&i2c->dev, "i2c trasfer error:%d", ret);
+   return 0;
+}
+
+static void set_pi3pdx_state(struct pi3dpx_priv *priv)
+{
+    priv->current_flip = priv->new_flip;
+	  switch (priv->typec_state) {
+    case TYPEC_DP_STATE_E:
+    case TYPEC_DP_STATE_C:
+      pi3dpx_i2c_write(priv, PI3DPX_BYTES_NUM, &redriver_config[priv->current_flip].full_dp_config);
+      break;
+    case TYPEC_DP_STATE_D:
+      pi3dpx_i2c_write(priv, PI3DPX_BYTES_NUM, &redriver_config[priv->current_flip].usb_dp_config);
+      break;
+    default:
+      pi3dpx_i2c_write(priv, PI3DPX_BYTES_NUM, &redriver_config[priv->current_flip].no_dp_config);
+  }
+  dev_info(priv->dev, "%s set flip:%d state:%ld\n", __func__, priv->current_flip, priv->typec_state);
+}
+
+static int pi3dpx_typec_mux_set(struct typec_mux_dev *mux,
+                struct typec_mux_state *state)
+{
+
+  struct pi3dpx_priv *priv = typec_mux_get_drvdata(mux);
+
+  unsigned long typec = state->mode;
+  if (typec != TYPEC_DP_STATE_E &&
+      typec != TYPEC_DP_STATE_C &&
+      typec != TYPEC_DP_STATE_D)
+    typec = TYPEC_STATE_SAFE;
+  if ((typec == priv->typec_state) && (priv->new_flip == priv->current_flip))
+    return 0;
+  mutex_lock(&priv->i2c_rw_lock);
+  priv->typec_state = typec;
+	set_pi3pdx_state(priv);
+  mutex_unlock(&priv->i2c_rw_lock);
+  return 0;
+}
+
+
+static int pi3dpx_orien_sw_set(struct typec_switch_dev *sw,
+            enum typec_orientation orien)
+{
+  struct pi3dpx_priv *priv = typec_switch_get_drvdata(sw);
+  if (orien == TYPEC_ORIENTATION_NONE)
+    return 0;
+  mutex_lock(&priv->i2c_rw_lock);
+  priv->new_flip = (orien == TYPEC_ORIENTATION_REVERSE) ? 1: 0;
+  mutex_unlock(&priv->i2c_rw_lock);
+	return 0;
+}
+static void pi3dpx_orien_sw_unregister(void *data)
+{
+	struct pi3dpx_priv *priv = data;
+	typec_switch_unregister(priv->sw);
+}
+
+static int pi3dpx_orien_sw_register(struct pi3dpx_priv *priv)
+{
+	struct typec_switch_desc sw_desc = {
+		.drvdata = priv,
+    .fwnode = dev_fwnode(priv->dev),
+    .set = pi3dpx_orien_sw_set,
+ 	};
+  priv->sw = typec_switch_register(priv->dev, &sw_desc);
+  if (IS_ERR(priv->sw)) {
+    dev_err(priv->dev, "Error register typec orientation switch: %ld\n",
+      PTR_ERR(priv->sw));
+    return PTR_ERR(priv->sw);
+  }
+  return devm_add_action_or_reset(priv->dev, pi3dpx_orien_sw_unregister, priv);
+}
+
+
+static void pi3dpx_typec_mux_unregister(void *data)
+{
+  struct pi3dpx_priv *priv = data;
+  typec_mux_unregister(priv->mux);
+}
+
+static int pi3dpx_typec_mux_register(struct pi3dpx_priv *priv)
+{
+  struct typec_mux_desc mux_desc = {
+    .drvdata = priv,
+    .fwnode = dev_fwnode(priv->dev),
+    .set = pi3dpx_typec_mux_set,
+  };
+  priv->mux = typec_mux_register(priv->dev, &mux_desc);
+  if (IS_ERR(priv->mux)) {
+    dev_err(priv->dev, "Error register typec mux: %ld\n",
+      PTR_ERR(priv->mux));
+    return PTR_ERR(priv->mux);
+  }
+  return devm_add_action_or_reset(priv->dev, pi3dpx_typec_mux_unregister, priv);
+}
+
+static int pi3dpx1207c_probe(struct i2c_client *i2c)
+{
+  	int ret;
+    struct pi3dpx_priv *priv;
+
+    priv = devm_kzalloc(&i2c->dev,
+                            sizeof(struct pi3dpx_priv), GFP_KERNEL);
+    if (priv == NULL)
+       return -ENOMEM;
+    mutex_init(&priv->i2c_rw_lock);
+    i2c_set_clientdata(i2c, priv);
+    priv->dev = &i2c->dev;
+    priv->typec_state = TYPEC_STATE_SAFE;
+    priv->client = i2c;
+    ret = pi3dpx_orien_sw_register(priv);
+    if (ret)
+      return ret;
+    ret = pi3dpx_typec_mux_register(priv);
+    if (ret)
+      return ret;
+    dev_info(priv->dev, "driver probed.\n");
+	return 0;
+}
+
+static const struct i2c_device_id pi3dpx1207c_id[] = {
+	{ "pi3dpx1207c", 0 },
+	{},
+};
+MODULE_DEVICE_TABLE(i2c, pi3dpx1207c_id);
+
+static const struct of_device_id pi3dpx1207c_dt_idtable[] = {
+	{ .compatible = "diodes,pi3dpx1207c" },
+	{},
+};
+MODULE_DEVICE_TABLE(of, pi3dpx1207c_dt_idtable);
+
+static struct i2c_driver pi3dpx1207c_driver = {
+	.driver		= {
+		.name	= "pi3dpx1207c",
+		.of_match_table	= pi3dpx1207c_dt_idtable,
+	},
+	.probe		= pi3dpx1207c_probe,
+	.id_table	= pi3dpx1207c_id,
+};
+
+module_i2c_driver(pi3dpx1207c_driver);
+
+MODULE_AUTHOR("Heiko Stuebner <heiko@sntech.de>");
+MODULE_DESCRIPTION("pi3dpx1207c redriver driver");
+MODULE_LICENSE("GPL");
Index: kernel-6_12/drivers/usb/typec/mux/Kconfig
===================================================================
--- kernel-6_12.orig/drivers/usb/typec/mux/Kconfig
+++ kernel-6_12/drivers/usb/typec/mux/Kconfig
@@ -25,6 +25,14 @@ config TYPEC_MUX_PI3USB30532
 	  Say Y or M if your system has a Pericom PI3USB30532 Type-C cross
 	  switch / mux chip found on some devices with a Type-C port.
 
+config TYPEC_MUX_PI3DPX1207C
+  tristate "Pericom PI3DPX1207C Type-C cross switch driver"
+  depends on I2C
+  help
+    Say Y or M if your system has a Pericom PI3DPX1207C Type-C cross
+    switch / mux chip provide DisplayPort alternate mode found on some
+    devices with a Type-C port.
+
 config TYPEC_MUX_INTEL_PMC
 	tristate "Intel PMC mux control"
 	depends on ACPI
Index: kernel-6_12/drivers/usb/typec/mux/Makefile
===================================================================
--- kernel-6_12.orig/drivers/usb/typec/mux/Makefile
+++ kernel-6_12/drivers/usb/typec/mux/Makefile
@@ -3,6 +3,7 @@
 obj-$(CONFIG_TYPEC_MUX_FSA4480)		+= fsa4480.o
 obj-$(CONFIG_TYPEC_MUX_GPIO_SBU)	+= gpio-sbu-mux.o
 obj-$(CONFIG_TYPEC_MUX_PI3USB30532)	+= pi3usb30532.o
+obj-$(CONFIG_TYPEC_MUX_PI3DPX1207C) += pi3dpx1207c.o
 obj-$(CONFIG_TYPEC_MUX_INTEL_PMC)	+= intel_pmc_mux.o
 obj-$(CONFIG_TYPEC_MUX_IT5205)		+= it5205.o
 obj-$(CONFIG_TYPEC_MUX_NB7VPQ904M)	+= nb7vpq904m.o
