Index: src/ash/components/arc/video_accelerator/gpu_arc_video_decode_accelerator.cc
===================================================================
--- src.orig/ash/components/arc/video_accelerator/gpu_arc_video_decode_accelerator.cc
+++ src/ash/components/arc/video_accelerator/gpu_arc_video_decode_accelerator.cc
@@ -386,7 +386,7 @@ void GpuArcVideoDecodeAccelerator::Initi
 
   profile_ = config->profile;
 #if BUILDFLAG(USE_CHROMEOS_MEDIA_ACCELERATION)
-  const bool use_vd = base::FeatureList::IsEnabled(arc::kVideoDecoder);
+  const bool use_vd = false;
 #if BUILDFLAG(USE_ARC_PROTECTED_MEDIA)
   if (!use_vd) {
     LOG(ERROR) << "Unsupported path: builds with USE_ARC_PROTECTED_MEDIA must "
Index: src/media/gpu/v4l2/v4l2_video_decode_accelerator.cc
===================================================================
--- src.orig/media/gpu/v4l2/v4l2_video_decode_accelerator.cc
+++ src/media/gpu/v4l2/v4l2_video_decode_accelerator.cc
@@ -413,7 +413,7 @@ void V4L2VideoDecodeAccelerator::AssignP
 
   enum v4l2_memory memory;
   if (!image_processor_device_ && output_mode_ == Config::OutputMode::IMPORT)
-    memory = V4L2_MEMORY_DMABUF;
+    memory = V4L2_MEMORY_MMAP;
   else
     memory = V4L2_MEMORY_MMAP;
 
@@ -1511,6 +1511,26 @@ bool V4L2VideoDecodeAccelerator::Dequeue
         return false;
       }
     } else {
+      if (!image_processor_device_ && output_mode_ == Config::OutputMode::IMPORT) {
+        const std::vector<base::ScopedFD>& fds = output_record.output_frame->DmabufFds();
+        size_t bytes_copyed = 0;
+        for (int i = 0; i < (int)fds.size(); i++) {
+          size_t plane_size = output_record.output_frame->layout().planes()[i].size;
+
+          uint8_t* dst = static_cast<uint8_t*> (mmap(NULL, plane_size,
+              PROT_READ | PROT_WRITE, MAP_SHARED, fds[i].get(), 
+              output_record.output_frame->layout().planes()[i].offset));
+          if (dst == MAP_FAILED) {
+            NOTIFY_ERROR(PLATFORM_FAILURE);
+            return false;
+          } else {
+            const uint8_t* data = static_cast<const uint8_t*>(buf->GetPlaneMapping(0)) + bytes_copyed;
+            memcpy(dst, data, plane_size);
+            bytes_copyed += plane_size;
+          }
+          munmap(dst, plane_size);
+        }
+      }
       SendBufferToClient(buf->BufferId(), bitstream_buffer_id, buf);
     }
   }
