diff --git a/rockchip.c b/rockchip.c
index 17478f3..19e371e 100644
--- a/rockchip.c
+++ b/rockchip.c
@@ -20,7 +20,7 @@
 #include "util.h"
 
 #define DRM_FORMAT_MOD_ROCKCHIP_AFBC                                                               \
-	DRM_FORMAT_MOD_ARM_AFBC(AFBC_FORMAT_MOD_BLOCK_SIZE_16x16 | AFBC_FORMAT_MOD_SPARSE |        \
+	DRM_FORMAT_MOD_ARM_AFBC(AFBC_FORMAT_MOD_BLOCK_SIZE_16x16 | /*AFBC_FORMAT_MOD_SPARSE |*/    \
 				AFBC_FORMAT_MOD_YTR)
 
 struct rockchip_private_map_data {
