Index: kernel-rockchip/mm/mprotect.c
===================================================================
--- kernel-rockchip.orig/mm/mprotect.c
+++ kernel-rockchip/mm/mprotect.c
@@ -713,8 +713,10 @@ static int do_mprotect_pkey(unsigned lon
 	mas_set(&mas, start);
 	vma = mas_find(&mas, ULONG_MAX);
 	error = -ENOMEM;
-	if (!vma)
+	if (!vma) {
+    pr_info("mprotect: vma:%p, pkey:%d\n", vma, pkey);
 		goto out;
+  }
 
 	if (unlikely(grows & PROT_GROWSDOWN)) {
 		if (vma->vm_start >= end)
@@ -803,6 +805,7 @@ static int do_mprotect_pkey(unsigned lon
 		vma = find_vma(current->mm, prev->vm_end);
 		if (!vma || vma->vm_start != nstart) {
 			error = -ENOMEM;
+      pr_info("mprotect:vma:%p, vm_start:0x%lx, nstart:0x%lx\n", vma, vma->vm_start, nstart);
 			break;
 		}
 		prot = reqprot;
@@ -816,7 +819,10 @@ out:
 SYSCALL_DEFINE3(mprotect, unsigned long, start, size_t, len,
 		unsigned long, prot)
 {
-	return do_mprotect_pkey(start, len, prot, -1);
+  int ret = do_mprotect_pkey(start, len, prot, -1);
+  if (ret)
+    pr_info("mprotect: start:0x%lx, len:%lu,prot:0x%lx, ret:%d\n", start, len, prot, ret);
+	return ret;
 }
 
 #ifdef CONFIG_ARCH_HAS_PKEYS
