Index: kernel-rockchip-6_1/drivers/gpu/drm/rockchip/dw-dp.c
===================================================================
--- kernel-rockchip-6_1.orig/drivers/gpu/drm/rockchip/dw-dp.c
+++ kernel-rockchip-6_1/drivers/gpu/drm/rockchip/dw-dp.c
@@ -3099,6 +3099,8 @@ static ssize_t dw_dp_aux_transfer(struct
 	u32 status, value;
 	ssize_t ret = 0;
 
+  if (!dw_dp_detect(dp))
+    return -EIO;
 	if (WARN_ON(msg->size > 16))
 		return -E2BIG;
 
