Index: kernel-rockchip/arch/arm64/boot/dts/rockchip/rk3588s-tablet-12c.dtsi
===================================================================
--- kernel-rockchip.orig/arch/arm64/boot/dts/rockchip/rk3588s-tablet-12c.dtsi
+++ kernel-rockchip/arch/arm64/boot/dts/rockchip/rk3588s-tablet-12c.dtsi
@@ -130,15 +130,13 @@
                 "Speaker", "LOUT2",
                 "Speaker", "ROUT2",
                 "Headphone", "Headphone Power",
-                "Headphone", "Headphone Power",
-                "Speaker", "Speaker Power",
                 "Speaker", "Speaker Power",
                 "LINPUT1", "Main Mic",
                 "LINPUT2", "Main Mic",
                 "RINPUT1", "Headset Mic",
                 "RINPUT2", "Headset Mic";
         pinctrl-names = "default";
-        pinctrl-0 = <&hp_det>;
+        pinctrl-0 = <&hp_det>, <&spk_con>, <&hp_con>;
         play-pause-key {
                 label = "playpause";
                 linux,code = <KEY_PLAYPAUSE>;
@@ -642,8 +640,10 @@
 		#sound-dai-cells = <0>;
 		compatible = "everest,es8388", "everest,es8323";
 		reg = <0x11>;
-		clocks = <&cru I2S0_8CH_MCLKOUT>;
+    clocks = <&mclkout_i2s0>;
 		clock-names = "mclk";
+    assigned-clocks = <&mclkout_i2s0>;
+    assigned-clock-rates = <12288000>;
 		pinctrl-names = "default";
 		pinctrl-0 = <&i2s0_mclk>;
 	};
@@ -723,6 +723,12 @@
         hp_det: hp-det {
             rockchip,pins = <1 RK_PC0 RK_FUNC_GPIO &pcfg_pull_none>;
         };
+        spk_con: spk-con {
+            rockchip,pins = <4 RK_PA5 RK_FUNC_GPIO &pcfg_pull_down>;
+        };
+        hp_con: hp-con {
+            rockchip,pins = <4 RK_PA4 RK_FUNC_GPIO &pcfg_pull_down>;
+        };
     };
 
 	hym8563 {
