Index: kernel-rockchip/arch/arm64/boot/dts/rockchip/rk3568-bpi-r2-pro.dts
===================================================================
--- kernel-rockchip.orig/arch/arm64/boot/dts/rockchip/rk3568-bpi-r2-pro.dts
+++ kernel-rockchip/arch/arm64/boot/dts/rockchip/rk3568-bpi-r2-pro.dts
@@ -181,17 +181,17 @@
 	};
 };
 
-&combphy0 {
+&combphy0_us {
 	/* used for USB3 */
 	status = "okay";
 };
 
-&combphy1 {
+&combphy1_us {
 	/* used for USB3 */
 	status = "okay";
 };
 
-&combphy2 {
+&combphy2_us {
 	/* used for SATA */
 	status = "okay";
 };
Index: kernel-rockchip/arch/arm64/boot/dts/rockchip/rk3568-evb1-v10.dts
===================================================================
--- kernel-rockchip.orig/arch/arm64/boot/dts/rockchip/rk3568-evb1-v10.dts
+++ kernel-rockchip/arch/arm64/boot/dts/rockchip/rk3568-evb1-v10.dts
@@ -160,11 +160,11 @@
 	};
 };
 
-&combphy0 {
+&combphy0_us {
 	status = "okay";
 };
 
-&combphy1 {
+&combphy1_us {
 	status = "okay";
 };
 
Index: kernel-rockchip/arch/arm64/boot/dts/rockchip/rk3568-rock-3a.dts
===================================================================
--- kernel-rockchip.orig/arch/arm64/boot/dts/rockchip/rk3568-rock-3a.dts
+++ kernel-rockchip/arch/arm64/boot/dts/rockchip/rk3568-rock-3a.dts
@@ -177,15 +177,15 @@
 	};
 };
 
-&combphy0 {
+&combphy0_us {
 	status = "okay";
 };
 
-&combphy1 {
+&combphy1_us {
 	status = "okay";
 };
 
-&combphy2 {
+&combphy2_us {
 	status = "okay";
 };
 
Index: kernel-rockchip/arch/arm64/boot/dts/rockchip/Makefile
===================================================================
--- kernel-rockchip.orig/arch/arm64/boot/dts/rockchip/Makefile
+++ kernel-rockchip/arch/arm64/boot/dts/rockchip/Makefile
@@ -66,8 +66,6 @@ dtb-$(CONFIG_ARCH_ROCKCHIP) += rk3588s-t
 dtb-$(CONFIG_ARCH_ROCKCHIP) += rk3588s-tablet-v11.dtb
 dtb-$(CONFIG_ARCH_ROCKCHIP) += rk3588s-tablet-12c-linux.dtb
 dtb-$(CONFIG_ARCH_ROCKCHIP) += rk3588-firefly-itx-3588j.dtb
-dtb-$(CONFIG_ARCH_ROCKCHIP) += rk3568-radxa-e25.dtb
-dtb-$(CONFIG_ARCH_ROCKCHIP) += rk3568-rock-3a.dtb
 dtb-$(CONFIG_ARCH_ROCKCHIP) += rk3588-blade3-v101-linux.dtb
 dtb-$(CONFIG_ARCH_ROCKCHIP) += rk3588-blueberry-edge-v10-linux.dtb
 dtb-$(CONFIG_ARCH_ROCKCHIP) += rk3588-blueberry-edge-v12-linux.dtb
Index: kernel-rockchip/arch/arm64/boot/dts/rockchip/rk3588-blueberry.dtsi
===================================================================
--- kernel-rockchip.orig/arch/arm64/boot/dts/rockchip/rk3588-blueberry.dtsi
+++ kernel-rockchip/arch/arm64/boot/dts/rockchip/rk3588-blueberry.dtsi
@@ -125,7 +125,7 @@
 };
 
 &display_subsystem {
-	clocks = <&hdptxphy_hdmi_clk0>, <&hdptxphy_hdmi_clk1>;
+	clocks = <&hdptxphy_hdmi0>, <&hdptxphy_hdmi1>;
 	clock-names = "hdmi0_phy_pll", "hdmi1_phy_pll";
 };
 
@@ -144,11 +144,11 @@
 	status = "okay";
 };
 
-&hdptxphy_hdmi_clk0 {
+&hdptxphy_hdmi0 {
 	status = "okay";
 };
 
-&hdptxphy_hdmi_clk1 {
+&hdptxphy_hdmi1 {
 	status = "okay";
 };
 
Index: kernel-rockchip/arch/arm64/boot/dts/rockchip/rk3588-nanopc-t6.dts
===================================================================
--- kernel-rockchip.orig/arch/arm64/boot/dts/rockchip/rk3588-nanopc-t6.dts
+++ kernel-rockchip/arch/arm64/boot/dts/rockchip/rk3588-nanopc-t6.dts
@@ -432,12 +432,12 @@
 	vpcie3v3-supply = <&vcc_3v3_pcie20>;
 	status = "okay";
 
-	pcie@20 {
+	pcie@0,0 {
 		reg = <0x00200000 0 0 0 0>;
 		#address-cells = <3>;
 		#size-cells = <2>;
 
-		r8125_u12: pcie@20,0 {
+		r8125_u12: pcie@0,0,0 {
 			reg = <0x000000 0 0 0 0>;
 			local-mac-address = [ 00 00 00 00 00 00 ];
 		};
@@ -456,12 +456,12 @@
 	vpcie3v3-supply = <&vcc_3v3_pcie20>;
 	status = "okay";
 
-	pcie@40 {
+	pcie@0,0 {
 		reg = <0x00400000 0 0 0 0>;
 		#address-cells = <3>;
 		#size-cells = <2>;
 
-		r8125_u10: pcie@40,0 {
+		r8125_u10: pcie@0,0,0 {
 			reg = <0x000000 0 0 0 0>;
 			local-mac-address = [ 00 00 00 00 00 00 ];
 		};
@@ -686,7 +686,7 @@
 
 
 &display_subsystem {
-	clocks = <&hdptxphy_hdmi_clk0>, <&hdptxphy_hdmi_clk1>;
+	clocks = <&hdptxphy_hdmi0>, <&hdptxphy_hdmi1>;
 	clock-names = "hdmi0_phy_pll", "hdmi1_phy_pll";
 };
 
Index: kernel-rockchip/arch/arm64/boot/dts/rockchip/rk3588-orangepi.dtsi
===================================================================
--- kernel-rockchip.orig/arch/arm64/boot/dts/rockchip/rk3588-orangepi.dtsi
+++ kernel-rockchip/arch/arm64/boot/dts/rockchip/rk3588-orangepi.dtsi
@@ -623,7 +623,7 @@
 
 
 &display_subsystem {
-	clocks = <&hdptxphy_hdmi_clk0>, <&hdptxphy_hdmi_clk1>;
+	clocks = <&hdptxphy_hdmi0>, <&hdptxphy_hdmi1>;
 	clock-names = "hdmi0_phy_pll", "hdmi1_phy_pll";
 };
 
Index: kernel-rockchip/arch/arm64/boot/dts/rockchip/rk3588-rock-5b.dts
===================================================================
--- kernel-rockchip.orig/arch/arm64/boot/dts/rockchip/rk3588-rock-5b.dts
+++ kernel-rockchip/arch/arm64/boot/dts/rockchip/rk3588-rock-5b.dts
@@ -633,7 +633,7 @@
 };
 
 &display_subsystem {
-	clocks = <&hdptxphy_hdmi_clk0>, <&hdptxphy_hdmi_clk1>;
+	clocks = <&hdptxphy_hdmi0>, <&hdptxphy_hdmi1>;
 	clock-names = "hdmi0_phy_pll", "hdmi1_phy_pll";
 
 	route {
@@ -657,11 +657,11 @@
 	};
 };
 
-&hdptxphy_hdmi_clk0 {
+&hdptxphy_hdmi0 {
 	status = "okay";
 };
 
-&hdptxphy_hdmi_clk1 {
+&hdptxphy_hdmi1 {
 	status = "okay";
 };
 
Index: kernel-rockchip/arch/arm64/boot/dts/rockchip/rk3588s-9tripod-linux.dts
===================================================================
--- kernel-rockchip.orig/arch/arm64/boot/dts/rockchip/rk3588s-9tripod-linux.dts
+++ kernel-rockchip/arch/arm64/boot/dts/rockchip/rk3588s-9tripod-linux.dts
@@ -357,7 +357,7 @@
 };
 
 &display_subsystem {
-    clocks = <&hdptxphy_hdmi_clk0>;
+    clocks = <&hdptxphy_hdmi0>;
     clock-names = "hdmi0_phy_pll";
 };
 
@@ -401,7 +401,7 @@
 		unprepare-delay-ms = <120>;
 		disable-delay-ms = <120>;
 		dsi,flags = <(MIPI_DSI_MODE_VIDEO | MIPI_DSI_MODE_VIDEO_BURST |
-			MIPI_DSI_MODE_LPM | MIPI_DSI_MODE_EOT_PACKET)>;
+			MIPI_DSI_MODE_LPM | MIPI_DSI_MODE_NO_EOT_PACKET)>;
 		dsi,format = <MIPI_DSI_FMT_RGB888>;
 		dsi,lanes  = <4>;
 
Index: kernel-rockchip/arch/arm64/boot/dts/rockchip/rk3588s-nanopi-r6-common.dtsi
===================================================================
--- kernel-rockchip.orig/arch/arm64/boot/dts/rockchip/rk3588s-nanopi-r6-common.dtsi
+++ kernel-rockchip/arch/arm64/boot/dts/rockchip/rk3588s-nanopi-r6-common.dtsi
@@ -694,6 +694,6 @@
 };
 
 &display_subsystem {
-	clocks = <&hdptxphy_hdmi_clk0>;
+	clocks = <&hdptxphy_hdmi0>;
 	clock-names = "hdmi0_phy_pll";
-};
\ No newline at end of file
+};
Index: kernel-rockchip/arch/arm64/boot/dts/rockchip/rk3588s-orangepi.dtsi
===================================================================
--- kernel-rockchip.orig/arch/arm64/boot/dts/rockchip/rk3588s-orangepi.dtsi
+++ kernel-rockchip/arch/arm64/boot/dts/rockchip/rk3588s-orangepi.dtsi
@@ -606,6 +606,6 @@
 
 /* Fix tty terminal out of screen, and most dclk of resolutions was not supported in hdmiphy clock from parent clock by default */
 &display_subsystem {
-	clocks = <&hdptxphy_hdmi_clk0>;
+	clocks = <&hdptxphy_hdmi0>;
 	clock-names = "hdmi0_phy_pll";
 };
Index: kernel-rockchip/arch/arm64/boot/dts/rockchip/rk3588s-radxa-cm5-io.dts
===================================================================
--- kernel-rockchip.orig/arch/arm64/boot/dts/rockchip/rk3588s-radxa-cm5-io.dts
+++ kernel-rockchip/arch/arm64/boot/dts/rockchip/rk3588s-radxa-cm5-io.dts
@@ -382,11 +382,11 @@
 };
 
 &display_subsystem {
-	clocks = <&hdptxphy_hdmi_clk0>;
+	clocks = <&hdptxphy_hdmi0>;
 	clock-names = "hdmi0_phy_pll";
 };
 
-&hdptxphy_hdmi_clk0 {
+&hdptxphy_hdmi0 {
 	status = "okay";
 };
 
Index: kernel-rockchip/arch/arm64/boot/dts/rockchip/rk3588s-rock-5a.dts
===================================================================
--- kernel-rockchip.orig/arch/arm64/boot/dts/rockchip/rk3588s-rock-5a.dts
+++ kernel-rockchip/arch/arm64/boot/dts/rockchip/rk3588s-rock-5a.dts
@@ -538,11 +538,11 @@
 };
 
 &display_subsystem {
-	clocks = <&hdptxphy_hdmi_clk0>;
+	clocks = <&hdptxphy_hdmi0>;
 	clock-names = "hdmi0_phy_pll";
 };
 
-&hdptxphy_hdmi_clk0 {
+&hdptxphy_hdmi0 {
 	status = "okay";
 };
 
Index: kernel-rockchip/arch/arm64/boot/dts/rockchip/rk3588s-khadas-edge2.dtsi
===================================================================
--- kernel-rockchip.orig/arch/arm64/boot/dts/rockchip/rk3588s-khadas-edge2.dtsi
+++ kernel-rockchip/arch/arm64/boot/dts/rockchip/rk3588s-khadas-edge2.dtsi
@@ -238,7 +238,7 @@
 		unprepare-delay-ms = <60>;
 		disable-delay-ms = <60>;
 		dsi,flags = <(MIPI_DSI_MODE_VIDEO | MIPI_DSI_MODE_VIDEO_BURST |
-			MIPI_DSI_MODE_LPM | MIPI_DSI_MODE_EOT_PACKET)>;
+			MIPI_DSI_MODE_LPM | MIPI_DSI_MODE_NO_EOT_PACKET)>;
 		dsi,format = <MIPI_DSI_FMT_RGB888>;
 		dsi,lanes  = <4>;
 		panel-init-sequence = [
@@ -877,7 +877,7 @@
 		unprepare-delay-ms = <60>;
 		disable-delay-ms = <60>;
 		dsi,flags = <(MIPI_DSI_MODE_VIDEO | MIPI_DSI_MODE_VIDEO_BURST |
-			MIPI_DSI_MODE_LPM | MIPI_DSI_MODE_EOT_PACKET)>;
+			MIPI_DSI_MODE_LPM | MIPI_DSI_MODE_NO_EOT_PACKET)>;
 		dsi,format = <MIPI_DSI_FMT_RGB888>;
 		dsi,lanes  = <4>;
 		panel-init-sequence = [
Index: kernel-rockchip/arch/arm64/boot/dts/rockchip/roc-rk3588s-pc-dual-mipi101-M101014-BE45-A1.dts
===================================================================
--- kernel-rockchip.orig/arch/arm64/boot/dts/rockchip/roc-rk3588s-pc-dual-mipi101-M101014-BE45-A1.dts
+++ kernel-rockchip/arch/arm64/boot/dts/rockchip/roc-rk3588s-pc-dual-mipi101-M101014-BE45-A1.dts
@@ -151,7 +151,7 @@
         mipi-data-delay-ms = <200>;
         size,width = <120>;
         size,height = <170>;
-        dsi,flags = <(MIPI_DSI_MODE_VIDEO | MIPI_DSI_MODE_VIDEO_BURST | MIPI_DSI_MODE_LPM | MIPI_DSI_MODE_EOT_PACKET)>;
+        dsi,flags = <(MIPI_DSI_MODE_VIDEO | MIPI_DSI_MODE_VIDEO_BURST | MIPI_DSI_MODE_LPM | MIPI_DSI_MODE_NO_EOT_PACKET)>;
         dsi,format = <MIPI_DSI_FMT_RGB888>;
         dsi,lanes  = <4>;
 
@@ -286,7 +286,7 @@
         mipi-data-delay-ms = <200>;
         size,width = <120>;
         size,height = <170>;
-        dsi,flags = <(MIPI_DSI_MODE_VIDEO | MIPI_DSI_MODE_VIDEO_BURST | MIPI_DSI_MODE_LPM | MIPI_DSI_MODE_EOT_PACKET)>;
+        dsi,flags = <(MIPI_DSI_MODE_VIDEO | MIPI_DSI_MODE_VIDEO_BURST | MIPI_DSI_MODE_LPM | MIPI_DSI_MODE_NO_EOT_PACKET)>;
         dsi,format = <MIPI_DSI_FMT_RGB888>;
         dsi,lanes  = <4>;
 
Index: kernel-rockchip/arch/arm64/boot/dts/rockchip/roc-rk3588s-pc-mipi101-M101014-BE45-A1.dts
===================================================================
--- kernel-rockchip.orig/arch/arm64/boot/dts/rockchip/roc-rk3588s-pc-mipi101-M101014-BE45-A1.dts
+++ kernel-rockchip/arch/arm64/boot/dts/rockchip/roc-rk3588s-pc-mipi101-M101014-BE45-A1.dts
@@ -159,7 +159,7 @@
         mipi-data-delay-ms = <200>;
         size,width = <120>;
         size,height = <170>;
-        dsi,flags = <(MIPI_DSI_MODE_VIDEO | MIPI_DSI_MODE_VIDEO_BURST | MIPI_DSI_MODE_LPM | MIPI_DSI_MODE_EOT_PACKET)>;
+        dsi,flags = <(MIPI_DSI_MODE_VIDEO | MIPI_DSI_MODE_VIDEO_BURST | MIPI_DSI_MODE_LPM | MIPI_DSI_MODE_NO_EOT_PACKET)>;
         dsi,format = <MIPI_DSI_FMT_RGB888>;
         dsi,lanes  = <4>;
 
@@ -296,7 +296,7 @@
         mipi-data-delay-ms = <200>;
         size,width = <120>;
         size,height = <170>;
-        dsi,flags = <(MIPI_DSI_MODE_VIDEO | MIPI_DSI_MODE_VIDEO_BURST | MIPI_DSI_MODE_LPM | MIPI_DSI_MODE_EOT_PACKET)>;
+        dsi,flags = <(MIPI_DSI_MODE_VIDEO | MIPI_DSI_MODE_VIDEO_BURST | MIPI_DSI_MODE_LPM | MIPI_DSI_MODE_NO_EOT_PACKET)>;
         dsi,format = <MIPI_DSI_FMT_RGB888>;
         dsi,lanes  = <4>;
 
Index: kernel-rockchip/arch/arm64/boot/dts/rockchip/rk3588-firefly-demo.dtsi
===================================================================
--- kernel-rockchip.orig/arch/arm64/boot/dts/rockchip/rk3588-firefly-demo.dtsi
+++ kernel-rockchip/arch/arm64/boot/dts/rockchip/rk3588-firefly-demo.dtsi
@@ -42,10 +42,10 @@
 };
 
 &spi1{
-    spi_demo: spi_demo@00{
+    spi_demo: spi_demo@0{
         compatible = "firefly,rk3588-spi";
         status = "disabled";
-        reg = <0x0>;
+        reg = <0>;
         spi-max-frequency = <50000000>;
         //spi-cpha;   /* SPI mode: CPHA=1 */
         //spi-cpol; 	/* SPI mode: CPOL=1 */
Index: kernel-rockchip/arch/arm64/boot/dts/rockchip/rk3588-firefly-itx-3588j.dtsi
===================================================================
--- kernel-rockchip.orig/arch/arm64/boot/dts/rockchip/rk3588-firefly-itx-3588j.dtsi
+++ kernel-rockchip/arch/arm64/boot/dts/rockchip/rk3588-firefly-itx-3588j.dtsi
@@ -180,10 +180,10 @@
 	pinctrl-0 = <&spi1m2_cs0 &spi1m2_pins>;
 	num-cs = <1>;
 	max-freq = <50000000>;
-    spidev1: spidev@00{
+    spidev1: spidev@0{
         compatible = "rockchip,spidev";
         status = "okay";
-        reg = <0x0>;
+        reg = <0>;
         spi-max-frequency = <50000000>;
     };
 };
Index: kernel-rockchip/arch/arm64/boot/dts/rockchip/rk3588s-nanopi-r6s.dts
===================================================================
--- kernel-rockchip.orig/arch/arm64/boot/dts/rockchip/rk3588s-nanopi-r6s.dts
+++ kernel-rockchip/arch/arm64/boot/dts/rockchip/rk3588s-nanopi-r6s.dts
@@ -79,12 +79,12 @@
 };
 
 &pcie2x1l1 {
-	pcie@30 {
+	pcie@0,0 {
 		reg = <0x00300000 0 0 0 0>;
 		#address-cells = <3>;
 		#size-cells = <2>;
 
-		r8125_u25: pcie@30,0 {
+		r8125_u25: pcie@0,0,0 {
 			reg = <0x000000 0 0 0 0>;
 			local-mac-address = [ 00 00 00 00 00 00 ];
 		};
@@ -92,12 +92,12 @@
 };
 
 &pcie2x1l2 {
-	pcie@40 {
+	pcie@1,0 {
 		reg = <0x00400000 0 0 0 0>;
 		#address-cells = <3>;
 		#size-cells = <2>;
 
-		r8125_u40: pcie@40,0 {
+		r8125_u40: pcie@1,0,0 {
 			reg = <0x000000 0 0 0 0>;
 			local-mac-address = [ 00 00 00 00 00 00 ];
 		};
Index: kernel-rockchip/drivers/input/remotectl/rockchip_pwm_remotectl.c
===================================================================
--- kernel-rockchip.orig/drivers/input/remotectl/rockchip_pwm_remotectl.c
+++ kernel-rockchip/drivers/input/remotectl/rockchip_pwm_remotectl.c
@@ -4,6 +4,7 @@
 #include <linux/io.h>
 #include <linux/irq.h>
 #include <linux/irqdesc.h>
+#include <linux/irqnr.h>
 #include <linux/irqdomain.h>
 #include <linux/module.h>
 #include <linux/of.h>
