Index: kernel-rockchip/drivers/hid/hid-multitouch.c
===================================================================
--- kernel-rockchip.orig/drivers/hid/hid-multitouch.c
+++ kernel-rockchip/drivers/hid/hid-multitouch.c
@@ -2143,6 +2143,10 @@ static const struct hid_device_id mt_dev
 	{ .driver_data = MT_CLS_SMART_TECH,
 		MT_USB_DEVICE(0x0b8c, 0x0092)},
 
+  /* fydetab duo touchpad */
+	{ .driver_data = MT_CLS_SMART_TECH,
+    MT_USB_DEVICE(0x05ac, 0x8502)},
+
 	/* Stantum panels */
 	{ .driver_data = MT_CLS_CONFIDENCE,
 		MT_USB_DEVICE(USB_VENDOR_ID_STANTUM_STM,
