Index: kernel-rockchip/arch/arm64/boot/dts/rockchip/rk3588-orangepi.dtsi
===================================================================
--- kernel-rockchip.orig/arch/arm64/boot/dts/rockchip/rk3588-orangepi.dtsi
+++ kernel-rockchip/arch/arm64/boot/dts/rockchip/rk3588-orangepi.dtsi
@@ -223,6 +223,10 @@
 	};
 };
 
+&av1d {
+  status = "okay";
+};
+
 &av1d_mmu {
 	status = "okay";
 };
Index: kernel-rockchip/arch/arm64/boot/dts/rockchip/rk3588s-khadas-edge2.dtsi
===================================================================
--- kernel-rockchip.orig/arch/arm64/boot/dts/rockchip/rk3588s-khadas-edge2.dtsi
+++ kernel-rockchip/arch/arm64/boot/dts/rockchip/rk3588s-khadas-edge2.dtsi
@@ -197,6 +197,10 @@
 	};
 };
 
+&av1d {
+  status = "okay";
+};
+
 &av1d_mmu {
 	status = "okay";
 };
Index: kernel-rockchip/arch/arm64/boot/dts/rockchip/rk3588s-rock-5a.dts
===================================================================
--- kernel-rockchip.orig/arch/arm64/boot/dts/rockchip/rk3588s-rock-5a.dts
+++ kernel-rockchip/arch/arm64/boot/dts/rockchip/rk3588s-rock-5a.dts
@@ -169,6 +169,10 @@
 	};
 };
 
+&av1d {
+  status = "okay";
+};
+
 &av1d_mmu {
 	status = "okay";
 };
