Index: libv4l-rkmpp/src/libv4l-rkmpp.c
===================================================================
--- libv4l-rkmpp.orig/src/libv4l-rkmpp.c
+++ libv4l-rkmpp/src/libv4l-rkmpp.c
@@ -961,6 +961,12 @@ static void *plugin_init(int fd)
 		goto err_free_ctx;
 	}
 
+	if (ctx->is_decoder) {
+		if (ctx->codecs)
+			free(ctx->codecs);
+		ctx->codecs = strdup("VP8:VP9:H.264:H.265:AV1");
+	}
+
 	/* Create eventfd to fake poll events */
 	ctx->eventfd = eventfd(0, EFD_NONBLOCK | EFD_CLOEXEC);
 	if (ctx->eventfd < 0) {
