Index: mpp/osal/mpp_allocator.c
===================================================================
--- mpp.orig/osal/mpp_allocator.c
+++ mpp/osal/mpp_allocator.c
@@ -148,17 +148,17 @@ MPP_RET mpp_allocator_get(MppAllocator *
             p->os_api = allocator_std;
         } break;
         case MPP_BUFFER_TYPE_ION : {
-            p->os_api = (mpp_rt_allcator_is_valid(MPP_BUFFER_TYPE_DMA_HEAP)) ? allocator_dma_heap :
-                        (mpp_rt_allcator_is_valid(MPP_BUFFER_TYPE_ION)) ? allocator_ion :
+            p->os_api = (mpp_rt_allcator_is_valid(MPP_BUFFER_TYPE_ION)) ? allocator_ion :
                         (mpp_rt_allcator_is_valid(MPP_BUFFER_TYPE_DRM)) ? allocator_drm :
+                        (mpp_rt_allcator_is_valid(MPP_BUFFER_TYPE_DMA_HEAP)) ? allocator_dma_heap :
                         allocator_std;
         } break;
         case MPP_BUFFER_TYPE_EXT_DMA: {
             p->os_api = allocator_ext_dma;
         } break;
         case MPP_BUFFER_TYPE_DRM : {
-            p->os_api = (mpp_rt_allcator_is_valid(MPP_BUFFER_TYPE_DMA_HEAP)) ? allocator_dma_heap :
-                        (mpp_rt_allcator_is_valid(MPP_BUFFER_TYPE_DRM)) ? allocator_drm :
+            p->os_api = (mpp_rt_allcator_is_valid(MPP_BUFFER_TYPE_DRM)) ? allocator_drm :
+                        (mpp_rt_allcator_is_valid(MPP_BUFFER_TYPE_DMA_HEAP)) ? allocator_dma_heap :
                         (mpp_rt_allcator_is_valid(MPP_BUFFER_TYPE_ION)) ? allocator_ion :
                         allocator_std;
         } break;
