Index: mpp/osal/driver/mpp_service.c
===================================================================
--- mpp.orig/osal/driver/mpp_service.c
+++ mpp/osal/driver/mpp_service.c
@@ -107,7 +107,7 @@ void check_mpp_service_cap(RK_U32 *codec
     RK_U32 i;
 
     /* for device check on startup */
-    mpp_env_get_u32("mpp_device_debug", &mpp_device_debug, 0);
+    mpp_env_get_u32("mpp_device_debug", &mpp_device_debug, 2);
 
     *codec_type = 0;
     memset(hw_ids, 0, sizeof(RK_U32) * 32);
Index: mpp/mpp/base/mpp_buffer_impl.c
===================================================================
--- mpp.orig/mpp/base/mpp_buffer_impl.c
+++ mpp/mpp/base/mpp_buffer_impl.c
@@ -1137,6 +1137,7 @@ static MppBufferGroupImpl *service_get_g
     flag = type_to_flag(type);
 
     p->flags = (MppAllocFlagType)flag;
+    mpp_err("get group: type:0x%x, mode:0x%x, buffer type:0x%x, flag:0x%x\n", type, mode, buffer_type, flag);
 
     {
         MppAllocator allocator = NULL;
Index: mpp/osal/mpp_platform.c
===================================================================
--- mpp.orig/osal/mpp_platform.c
+++ mpp/osal/mpp_platform.c
@@ -118,7 +118,7 @@ static void mpp_plat_srv_init()
     cap->poll_cmd = MPP_CMD_POLL_BASE + 1;
     cap->ctrl_cmd = MPP_CMD_CONTROL_BASE + 0;
 
-    mpp_env_get_u32("mpp_debug", &mpp_debug, 0);
+    mpp_env_get_u32("mpp_debug", &mpp_debug, 0x10);
 
     /* read soc name */
     srv->soc_name = mpp_get_soc_name();
Index: mpp/osal/mpp_runtime.c
===================================================================
--- mpp.orig/osal/mpp_runtime.c
+++ mpp/osal/mpp_runtime.c
@@ -70,7 +70,7 @@ static void mpp_rt_srv_init()
 {
     MppRuntimeSrv *srv = srv_runtime;
 
-    mpp_env_get_u32("mpp_rt_debug", &mpp_rt_debug, 0);
+    mpp_env_get_u32("mpp_rt_debug", &mpp_rt_debug, 1);
 
     if (srv)
         return;
Index: mpp/osal/driver/vcodec_service.c
===================================================================
--- mpp.orig/osal/driver/vcodec_service.c
+++ mpp/osal/driver/vcodec_service.c
@@ -539,11 +539,16 @@ MPP_RET vcodec_service_init(void *ctx, M
         if (!ret) {
             vcodec_ioctl_version = 0;
         } else {
+            mpp_err("failed to ioctl VPU_IOC_SET_CLIENT_TYPE:%lu, name:%s, type:%u, ret:%d MPP_OK:%d\n", VPU_IOC_SET_CLIENT_TYPE, name, client_type, ret, MPP_OK);
             ret = (RK_S32)ioctl(p->fd, VPU_IOC_SET_CLIENT_TYPE_U32, (RK_U32)client_type);
-            if (!ret)
+            if (!ret) {
                 vcodec_ioctl_version = 1;
+            } else {
+              mpp_err("failed to ioctl VPU_IOC_SET_CLIENT_TYPE_U32:%lu, name:%s, type:%u, ret:%d MPP_OK:%d\n", VPU_IOC_SET_CLIENT_TYPE_U32, name, client_type, ret, MPP_OK);
+            }
         }
-        mpp_assert(ret == MPP_OK);
+      //  mpp_assert(ret == MPP_OK);
+        ret = 0;
     } else {
         RK_U32 cmd = (vcodec_ioctl_version == 0) ?
                      (VPU_IOC_SET_CLIENT_TYPE) :
Index: mpp/osal/mpp_log.c
===================================================================
--- mpp.orig/osal/mpp_log.c
+++ mpp/osal/mpp_log.c
@@ -20,7 +20,7 @@
 extern "C" {
 #endif
 
-RK_U32 mpp_debug = 0;
+RK_U32 mpp_debug = MPP_DBG_INFO;
 
 // TODO: add log timing information and switch flag
 static const char *msg_log_warning = "log message is long\n";
