diff --git a/osal/allocator/allocator_drm.c b/osal/allocator/allocator_drm.c
index c0bb9f256bfb..41eb831aa5ad 100644
--- a/osal/allocator/allocator_drm.c
+++ b/osal/allocator/allocator_drm.c
@@ -51,7 +51,7 @@ typedef struct {
     RK_U32  flags;
 } allocator_ctx_drm;
 
-static const char *dev_drm = "/dev/dri/card0";
+static const char *dev_drm = "/dev/dri/renderD128";
 
 static int drm_ioctl(int fd, int req, void *arg)
 {
