Index: mesa/src/panfrost/vulkan/panvk_physical_device.c
===================================================================
--- mesa.orig/src/panfrost/vulkan/panvk_physical_device.c
+++ mesa/src/panfrost/vulkan/panvk_physical_device.c
@@ -1006,15 +1006,16 @@ panvk_physical_device_init(struct panvk_
    switch (arch) {
    case 6:
    case 7:
-      if (!getenv("PAN_I_WANT_A_BROKEN_VULKAN_DRIVER")) {
-         result = panvk_errorf(instance, VK_ERROR_INCOMPATIBLE_DRIVER,
-                               "WARNING: panvk is not well-tested on v%d, "
-                               "pass PAN_I_WANT_A_BROKEN_VULKAN_DRIVER=1 "
-                               "if you know what you're doing.", arch);
-         goto fail;
-      }
+/* remove crashing warning.
+//      if (!getenv("PAN_I_WANT_A_BROKEN_VULKAN_DRIVER")) {
+//         result = panvk_errorf(instance, VK_ERROR_INCOMPATIBLE_DRIVER,
+//                               "WARNING: panvk is not well-tested on v%d, "
+//                               "pass PAN_I_WANT_A_BROKEN_VULKAN_DRIVER=1 "
+//                               "if you know what you're doing.", arch);
+//         goto fail;
+//      }
+*/
       break;
-
    case 10:
       break;
 
